File: INFORMÁTICA/PROGRAMACIÓN Y DISEÑO ORIENTADO A OBJETOS/PRÁCTICAS/PROYECTOS/EXÁMENES/examen_parcial_18-19/Deepspace/SpaceCity.rb
#encoding: UTF-8

require_relative 'SpaceStation'
require_relative 'Transformation'
require_relative 'SpaceCityToUI'

module Deepspace

    #Encapsula la mejora de una Space Station a SpaceCity.
    class SpaceCity < SpaceStation

        #base: Estacion espacial principal de la SpaceCity
        #collaborators: Resto de estaciones espaciales que componen la SpaceCity.
        def initialize(base, rest)
            newCopy(base)

            @base=base
            @collaborators=Array.new(rest)
        end

        attr_reader :collaborators

        #Se suma la potencia de disparo de todas las estaciones de la Space City.
        #Override
        def fire()
        
            fire= super

            @collaborators.each do |collaborator|
                fire+=collaborator.fire()
            end
            
            return fire

        end

        #Se suma la potencia de escudo de todas las estaciones de la Space City.
        #Override
        def protection()
        
            protection= super
            
            @collaborators.each do |collaborator|
                protection+=collaborator.protection()
            end

            return protection

        end


        #Se acepta un Loot pero se anula la posibilidad de cualquier otra 
        #transformacion.
        #Override
        def setLoot(loot)
            super
            return Transformation::NOTRANSFORM
        
        end

        #Override
        def getUIversion
            return (SpaceCityToUI.new(self))
        end
    end
end

File: INFORMÁTICA/PROGRAMACIÓN Y DISEÑO ORIENTADO A OBJETOS/PRÁCTICAS/PROYECTOS/EXÁMENES/examen_2_19-20_definitivo/deepspace/src/deepspace/SpaceCity.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author leandrofdez
 */

//Encapsula la mejora de una Space Station a SpaceCity.
public class SpaceCity extends SpaceStation {
    
    //Estacion espacial principal de la SpaceCity
    private SpaceStation base;
    
    //Resto de estaciones espaciales que componen la SpaceCity.
    private ArrayList<SpaceStation> collaborators;
    
    
    
    
    public SpaceCity(SpaceStation base, ArrayList<SpaceStation> rest){
        super(base);
        this.base= base;
        this.collaborators= new ArrayList<>(rest);
    }
    
    
    
    //Getters
    public ArrayList<SpaceStation> getCollaborators(){
        
        return this.collaborators;
    }
    
    @Override
    public SpaceCityToUI getUIversion(){return (new SpaceCityToUI(this));}
    
    
    /*
    Se suma la potencia de disparo de todas las estaciones de la Space City.
    */
    @Override
    public float fire(){
        
        float fire= super.fire();
        
        for (int i=0; i<collaborators.size();i++) 
            fire+=collaborators.get(i).fire();
        
        return fire;
        
        
    }
    
    
    /*
    Se suma la potencia de escudo de todas las estaciones de la Space City.
    */
    @Override
    public float protection(){
        
        float protection= super.protection();
        
        for (int i=0; i<collaborators.size();i++) 
            protection+=collaborators.get(i).protection();
        
        return protection;
        
        
    }
    
    /*
    Se acepta un Loot pero se anula la posibilidad de cualquier otra 
    transformacion.
    */
    @Override
    public Transformation setLoot(Loot loot){
        
        super.setLoot(loot);
        return Transformation.NOTRANSFORM;
        
    }
    
}
